### ISSUE TITLE
ENH: Modify signal.butter to give an option to satisfy stop band or pass band attenuation depending on the user requirements.

### ISSUE THREAD

**Issue by Adarsh-NP** [scipy/scipy#15430]
### Is your feature request related to a problem? Please describe.

Buttord in scipy.signal, for calculating the cut off frequency for the low pass prototype, currently satisfies the pass band attenuation and over-satisfies the stop band attenuation specifications by default. Since Matlab does the same automatically so it makes sense to do so however, having an option to choose whether to satisfy the stop band attenuation or to satisfy pass band attenuation makes sense.

### Describe the solution you'd like.

The snippet where the cut off frequency of low pass prototype is calculated can be modified with if else conditions to satisfy this. 

The function prototype of buttord can be modified to receive one additional optional argument, which if provided would satisfy the stop band attenuation correctly and over satisfy the pass band attenuation.

`def buttord(wp, ws, gpass, gstop, analog=False, fs=None)` can be modified to
`def buttord(wp, ws, gpass, gstop, analog=False, fs=None, atype=1)` where atype refers to the attenuation type (pass band or stop band).

And in the snippet where we calculate the cut off frequency of the low pass prototype can be modified to:
`
if not atype:
    try:
        W0 = (GPASS - 1.0) ** (-1.0 / (2.0 * ord))
    except ZeroDivisionError:
        W0 = 1.0
        warnings.warn("Order is zero...check input parameters.",
                      RuntimeWarning, 2)
else:
    try:
          W0 = (GSTOP - 1.0) ** (-1.0 / (2.0 * ord))
      except ZeroDivisionError:
          W0 = 1.0
          warnings.warn("Order is zero...check input parameters.",
                        RuntimeWarning, 2)
`

@endolith @rgommers would like to know what you think of this minor yet useful enhancement.

**Comment by endolith** [comment]
I agree that this would be useful, but not sure what the best API is. It should probably be independent for both ends? and should apply to all the *ord functions.

**Comment by Adarsh-NP** [comment]
@endolith, as far as I have studied in my DSP course in undergrad this semester, I think this choice is for sure there in the cheby and butter filters. For other filters, I am not aware of but I am willing to do research on it. 

The general questions that I have come across in filter design mention whether we need to specify the pass band ripples or the stop band ripples correctly and same with butter so that's where this idea came to me.

Regarding API design, I feel currently we can go ahead with modifying function prototype to accommodate the atype (attenuation type) variable  as an optional parameter as this does not break anything I think but I feel the maintainers and core contributors should have a discussion that follows the contribution guidelines as I am new to scipy.

Also one query, I did not understand what you meant by independent from both ends. It would be kind of you to elaborate on that once.

**Comment by endolith** [comment]
Maybe I'm misunderstanding what you mean. I'm referring to a case like the example:

![image](https://user-images.githubusercontent.com/58611/149833477-37ef459c-c060-4e9c-8c3c-60d7cd76563a.png)

The gray boxes represent the constraints (`[20, 50], [14, 60]`), and a filter can be designed that fits anywhere within those constraints.  Currently it designs the narrowest filter that satisfies the constraints, but sometimes you might want the widest filter that satisfies the constraints, and which type of optimization you want might vary for the HPF end vs the LPF end.

and the same concerns apply to the other functions like `cheb2ord`:

![image](https://user-images.githubusercontent.com/58611/149833700-cc2bb35a-77ff-4526-9bca-ae6cbc2cb915.png)

**Comment by Adarsh-NP** [comment]
Yes, @endolith what you're saying is what we get when we vary the cutoff frequency of the analog low pass prototype filter that we design in the initial step. That gives some control. Basically the choices are:

A. satisfying the pass band attenuation and over satisfying the stop band attenuation (attenuation at least the value specified) which is what our current code does.
B. satisfying the stop band and over-satisfying the stop band (here it means that the passband attenuation will be less than the value specified).

These two would result in different cut off frequencies of our analog low pass prototype which will directly affect the width of the filter.

Now I'm sure this behaviour is tunable with the Butterworth filters where we can choose option A or B and do the necessary substitution. I think this should be the case with other *ord filters too but I'm not sure if it's tunable in other *ord filters so I'll have to read a bit on it. 

But overall including this feature is a good enhancement to our functionality.
